zapcore/encoder_test.go

```diff
@@ -218,97 +257,59 @@ func TestJSONEncoderConfiguration(t *testing.T) {
 				EncodeDuration: base.EncodeDuration,
 				EncodeLevel:    func(l Level, enc ArrayEncoder) { enc.AppendString(strings.ToUpper(l.String())) },
 			},
-			expected: `{"L":"INFO","T":0,"N":"main","C":"foo.go:42","M":"hello","S":"fake-stack"}`,
+			expectedJSON:    `{"L":"INFO","T":0,"N":"main","C":"foo.go:42","M":"hello","S":"fake-stack"}`,
+			expectedConsole: "0\tINFO\tmain@foo.go:42\thello\nfake-stack",
+		},
+		{
+			desc: "close all open namespaces",
+			cfg: EncoderConfig{
+				LevelKey:       "L",
+				TimeKey:        "T",
+				MessageKey:     "M",
+				NameKey:        "N",
+				CallerKey:      "C",
+				StacktraceKey:  "S",
+				EncodeTime:     base.EncodeTime,
+				EncodeDuration: base.EncodeDuration,
+				EncodeLevel:    base.EncodeLevel,
+			},
+			extra: func(enc Encoder) {
+				enc.OpenNamespace("outer")
+				enc.OpenNamespace("inner")
+				enc.AddString("foo", "bar")
+				enc.OpenNamespace("innermost")
+			},
+			expectedJSON: `{"L":"info","T":0,"N":"main","C":"foo.go:42","M":"hello","outer":{"inner":{"foo":"bar","innermost":{}}},"S":"fake-stack"}`,
+			expectedConsole: "0\tinfo\tmain@foo.go:42\thello\t" +
+				`{"outer": {"inner": {"foo": "bar", "innermost": {}}}}` +
+				"\nfake-stack",
 		},
 	}
 
 	for i, tt := range tests {
-		enc := NewJSONEncoder(tt.cfg)
+		json := NewJSONEncoder(tt.cfg)
+		console := NewConsoleEncoder(tt.cfg)
 		if tt.extra != nil {
-			tt.extra(enc)
+			tt.extra(json)
+			tt.extra(console)
+		}
+		jsonOut, jsonErr := json.EncodeEntry(_testEntry, nil)
+		if assert.NoError(t, jsonErr, "Unexpected error JSON-encoding entry in case #%d.", i) {
+			assert.Equal(
+				t,
+				tt.expectedJSON+"\n",
+				string(jsonOut),
+				"Unexpected JSON output: expected to %v.", tt.desc,
+			)
 		}
-		buf, err := enc.EncodeEntry(entry, nil)
-		if assert.NoError(t, err, "Unexpected error encoding entry in case #%d.", i) {
-			assert.Equal(t, tt.expected+"\n", string(buf), "Unexpected output: expected to %v.", tt.desc)
+		consoleOut, consoleErr := console.EncodeEntry(_testEntry, nil)
+		if assert.NoError(t, consoleErr, "Unexpected error console-encoding entry in case #%d.", i) {
+			assert.Equal(
+				t,
+				tt.expectedConsole+"\n",
+				string(consoleOut),
+				"Unexpected console output: expected to %v.", tt.desc,
+			)
 		}
 	}
 }
-
-func TestJSONEncodeEntry(t *testing.T) {
-	epoch := time.Unix(0, 0)
-	withJSONEncoder(func(enc Encoder) {
-		// Messages should be escaped.
-		enc.AddString("foo", "bar")
-		buf, err := enc.EncodeEntry(Entry{
-			LoggerName: "",
-			Level:      InfoLevel,
-			Message:    `hello\`,
-			Time:       epoch,
-		}, nil)
-		assert.NoError(t, err, "EncodeEntry returned an unexpected error.")
-		assert.Equal(
-			t,
-			`{"level":"info","ts":0,"msg":"hello\\","foo":"bar"}`+"\n",
```

Comment:
Escaping in general is, but I'll amend the test above to make sure we're also escaping the message.